### PR TITLE
Implement behavior of legacy larg4 module

### PR DESCRIPTION
**Pull request by hanswenzel**
Implement behavior of legacy larg4 module: 
which modified the trackID stored in the SimEnergyDeposit. If the particle
with that trackID was to be dropped, the value stored in the EnergyDeposit
was -1 times the value of the parent that was retained in the collection.

This is the default but a parameter of the larg4::LArG4DetectorService
updateSimEnergyDeposits_{p.get<bool>("UpdateSimEnergyDeposits", true)}
can be set to false then the SimEnergyDeposits are not modified.
(This is actually the prefered behavior). One can always relate the
SimEnergyDeposit to the ParentID by using the provided

   std::unique_ptr<std::map<int,std::set<int>>> DroppedTracksCollection()
where the key is the TrackID of the parent, the set are the trackid for which
no MCParticle was created.

This requires an updated version of: LArSoft/lardataobj

### CODE REVIEW

**Comment by lgarren** [comment]
trigger build with pull requests #38, LArSoft/lardataobj#21

File: larg4/Analysis/CheckSimEnergyDeposit_module.cc [hunk 1]
```diff
@@ -77,17 +75,8 @@ void larg4::CheckSimEnergyDeposit::analyze(const art::Event& event)
...
-        hit.GetEdep(),
-        hit.GetEdepEM(),
-        hit.GetEdepnonEM(),
-        hit.GetXpos(),
-        hit.GetYpos(),
-        hit.GetZpos(),
-        hit.GetTime());
-      */
     }
+    std::cout<< std::endl;
```
Comment (knoepfel) [inline larg4/Analysis/CheckSimEnergyDeposit_module.cc]: Remove this line.

File: larg4/pluginActions/ParticleListAction_service.cc [hunk 2]
```diff
@@ -210,6 +206,26 @@ namespace larg4 {
...
+    auto itr = fTargetIDMap.find(trackid);
+    if(itr == fTargetIDMap.end()){
+      mf::LogWarning("ParticleListActionService::getStorableTrackID")
+	<< "Track ID " << trackid << " not found in the TargetIDMap.  "
+	<< "Returning original trackid, but there may be issues with truth-matching.";
+      return trackid;
+    }
+    return itr->second;
+  }
+  */
```
Comment (knoepfel) [inline larg4/pluginActions/ParticleListAction_service.cc]: Suggest removing this commented-out code.

File: larg4/pluginActions/ParticleListAction_service.cc [hunk 3]
```diff
@@ -351,9 +362,17 @@ namespace larg4 {
...
           // clear current particle as we are not stepping this particle and
           // adding trajectory points to it
-	  fdroppedTracksSet.insert(trackID);
+	  if (fdroppedTracksMap.find(this->GetParentage(trackID))==fdroppedTracksMap.end())
+	    {
+	      fdroppedTracksMap[this->GetParentage(trackID)]={trackID};
+	    } else
+	    {
+	      fdroppedTracksMap[this->GetParentage(trackID)].insert(trackID);
+	    }
```
Comment (knoepfel) [inline larg4/pluginActions/ParticleListAction_service.cc]: This code can be achieved equivalently by just typing:

```c++
fdroppedTracksMap[this->GetParentage(trackID)].insert(trackID);
```

If the `std::set` does not exist for `this-GetParentage(trackID)`, it will be created.  Otherwise, the already-stored `std::set` will be returned by calling `std::map::operator[]`.

File: larg4/pluginActions/ParticleListAction_service.cc [hunk 4]
```diff
@@ -363,13 +382,19 @@ namespace larg4 {
...
       G4double energy = track->GetKineticEnergy();
       if (energy < fenergyCut) {
-	fdroppedTracksSet.insert(trackID);
+	  if (fdroppedTracksMap.find(this->GetParentage(trackID))==fdroppedTracksMap.end())
+	    {
+	      fdroppedTracksMap[this->GetParentage(trackID)]={trackID};
+	    } else
+	    {
+	      fdroppedTracksMap[this->GetParentage(trackID)].insert(trackID);
+	    }
```
Comment (knoepfel) [inline larg4/pluginActions/ParticleListAction_service.cc]: Suggest:

```c++
fdroppedTracksMap[this->GetParentage(trackID)].insert(trackID);
```

**Review by knoepfel — COMMENTED** [review]
Thanks for the PR, Hans.  Just a few comments below.

File: CMakeLists.txt [hunk 5]
```diff
@@ -8,7 +8,7 @@
 
 cmake_minimum_required(VERSION 3.19 FATAL_ERROR)
 
-project(larg4 VERSION 09.08.00 LANGUAGES CXX)
+project(larg4 VERSION 09.09.00 LANGUAGES CXX)
```
Comment (knoepfel) [inline CMakeLists.txt]: Revert this change as well--Lynn will make it when she's ready to tag the package.

Comment (lgarren) [reply CMakeLists.txt]: Note that updating a package version number in a PR may have unintended consequences when running the CI.

File: larg4/Analysis/CheckSimEnergyDeposit_module.cc (hunk 1 above)
Comment (hanswenzel) [reply larg4/Analysis/CheckSimEnergyDeposit_module.cc]: done

File: larg4/pluginActions/ParticleListAction_service.cc (hunk 2 above)
Comment (hanswenzel) [reply larg4/pluginActions/ParticleListAction_service.cc]: done

File: larg4/pluginActions/ParticleListAction_service.cc (hunk 3 above)
Comment (hanswenzel) [reply larg4/pluginActions/ParticleListAction_service.cc]: done

File: larg4/pluginActions/ParticleListAction_service.cc (hunk 4 above)
Comment (hanswenzel) [reply larg4/pluginActions/ParticleListAction_service.cc]: done

File: CMakeLists.txt (hunk 5 above)
Comment (hanswenzel) [reply CMakeLists.txt]: done

**Comment by lgarren** [comment]
trigger build with pull requests #39, LArSoft/lardataobj#21

**Comment by lgarren** [comment]
@LArSoft/level-2-managers  would you please take a look at this PR, along with the companion PR https://github.com/LArSoft/lardataobj/pull/21, and approve or comment?

**Comment by absolution1** [comment]
approve

**Comment by lgarren** [comment]
approve
